### PR TITLE
Configure Cassandra nodes to lookup their IP address by their fully qualified domain name

### PR DESCRIPTION
**Pull request by wallrj**
* Also ensure that they only ever use the Kubernetes seed provider service
* And ensure that that service publishes the seed IP addresses as soon as they are known,
* so that seeds them selves can find themselves when they are starting up.

This is a cut down version of #330 
* without the headless service per statefulset, which I think may be unnecessary, since nodes only have to be able to lookup their own FQDN.
* and without the E2E tests, because I haven't yet figured out how to force minikube to change the IP address of a pod when it's recreated.

Fixes: #319

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by munnerz** [comment]
> and without the E2E tests, because I haven't yet figured out how to force minikube to change the IP address of a pod when it's recreated.

When a pod is deleted, in 99% of cases it should have a new IP address. Can we make a test that deletes the pod and verifies the IP has changed? (perhaps we delete the pod again if it hasn't changed first time?).

We definitely need an e2e test for this - ideally the test would be labelled "Resiliency" too, as the first of a suite of resiliency tests.

File: pkg/cassandra/nodetool/nodetool.go [hunk 1]
```diff
@@ -135,17 +124,6 @@ func (t *tool) Status() (NodeMap, error) {
...
-		return nil, fmt.Errorf(
-			"mapped nodes must be a superset of leaving, joining and moving nodes. "+
-				"Leaving: %v, "+
-				"Joining: %v, "+
-				"Moving: %v, "+
-				"Mapped: %v",
-			leavingNodes, joiningNodes, movingNodes, mappedNodes,
-		)
-	}
-
```
Comment (munnerz) [inline pkg/cassandra/nodetool/nodetool.go]: I'm confused - isn't this meant to be a part of another PR?

If yes and this is just stacked, it'd be great if we can stop stacking PRs too much as it makes it far harder to know which to review first. If we keep them separate, PRs that depend on [some other] PR will fail and/or we can add `/hold` and comment to inform a reviewer that "this PR depends on PR xyz being merged first"

File: pkg/controllers/cassandra/nodepool/resource.go [hunk 2]
```diff
@@ -194,6 +195,34 @@ func StatefulSetForCluster(
...
+									Value: " ",
+								},
+								{
+									Name:  "CASSANDRA_BROADCAST_ADDRESS",
+									Value: " ",
+								},
+								{
+									Name:  "CASSANDRA_RPC_ADDRESS",
+									Value: " ",
+								},
```
Comment (munnerz) [inline pkg/controllers/cassandra/nodepool/resource.go]: Do we use the Cassandra RPC interface, and do we want to listen on the default pod IP for RPC connections? It seems to me like we'd only want to listen on 127.0.0.1?

File: pkg/controllers/cassandra/nodepool/resource.go [hunk 3]
```diff
@@ -194,6 +195,34 @@ func StatefulSetForCluster(
...
+								// Set a non-existent default seed.
+								// The Kubernetes Seed Provider will fall back to a default seed host if it can't look up seeds via the CASSANDRA_SERVICE.
+								// And if the CASSANDRA_SEEDS environment variable is not set, it defaults to localhost.
+								// Which could cause confusion if a non-seed node is temporarily unable to lookup the seed nodes from the service.
```
Comment (munnerz) [inline pkg/controllers/cassandra/nodepool/resource.go]: What would happen if we were to set this to localhost instead? Would the node fail to boot, or would it cause a split brain of some sort?

File: pkg/controllers/cassandra/nodepool/resource.go [hunk 4]
```diff
@@ -194,6 +195,34 @@ func StatefulSetForCluster(
...
+								// We want the list of seeds to be strictly controlled by the service.
+								// See:
+								// https://github.com/docker-library/cassandra/blame/master/3.11/docker-entrypoint.sh#L31 and
+								// https://github.com/apache/cassandra/blob/cassandra-3.11.2/conf/cassandra.yaml#L416 and
+								// https://github.com/kubernetes/examples/blob/cabf8b8e4739e576837111e156763d19a64a3591/cassandra/java/src/main/java/io/k8s/cassandra/KubernetesSeedProvider.java#L69 and
+								// https://github.com/kubernetes/examples/blob/cabf8b8e4739e576837111e156763d19a64a3591/cassandra/go/main.go#L51
+								{
+									Name:  "CASSANDRA_SEEDS",
+									Value: "black-hole-dns-name",
```
Comment (munnerz) [inline pkg/controllers/cassandra/nodepool/resource.go]: I'm slightly concerned this could become an attack vector. What if someone creates a service named black-hole-dns-name? Could an attacker gain access to the contents of the Cassandra DB as a result?

If possible, I'd rather set this to a value that causes Cassandra to fail hard/crash (as this value *should* only be referenced in the event of a DNS resolution failure)

Comment (wallrj) [reply pkg/controllers/cassandra/nodepool/resource.go]: I think the answer is to remove the fallback code from the Kubernetes Seed Provider.
If it can't lookup the seeds from the Kubernetes it should log an error and return an empty list,
rather than returning a default.

File: pkg/controllers/cassandra/nodepool/resource.go (hunk 2 above)
Comment (wallrj) [reply pkg/controllers/cassandra/nodepool/resource.go]: We don't use it, but I'm not sure whether Cassandra nodes talk to each other with this protocol.
The default is to listen on all addresses:
 * https://github.com/docker-library/cassandra/blob/master/3.11/docker-entrypoint.sh#L29

So setting it here may not be ideal, but I can improve or remove it later.

Comment (munnerz) [reply pkg/controllers/cassandra/nodepool/resource.go]: 👍

File: pkg/controllers/cassandra/nodepool/resource.go (hunk 3 above)
Comment (wallrj) [reply pkg/controllers/cassandra/nodepool/resource.go]: I think it would mean that new nodes might consider themselves seeds in the event that the KubernetesSeedProvider plugin encounters an error.
So yeah, a split brain situation I think.

File: pkg/controllers/cassandra/nodepool/resource.go (hunk 4 above)
Comment (wallrj) [reply pkg/controllers/cassandra/nodepool/resource.go]: I'll do this in a follow up branch, where I remove the reliance on the Docker image entrypoint and instead create a Navigator default config.
